### PR TITLE
add apoc.convert.toInteger, apoc.convert.toDouble functions - #593

### PR DESCRIPTION
**Pull request by zygm0nt**

### CODE REVIEW

File: src/main/java/apoc/convert/Convert.java [hunk 1]
```diff
@@ -163,5 +165,85 @@ else if (list.getClass().isArray() && !list.getClass().getComponentType().isPrim
...
+        if (obj == null || obj.equals("")) {
+            return null;
+        }
+
+        Types varType = Types.of(obj);
+        Number result;
+        switch (varType) {
+            case INTEGER:
+            case FLOAT:
+                result = ((Number) obj).intValue();
```
Comment (jexp) [inline src/main/java/apoc/convert/Convert.java]: Cypher only knows long (which it calls Integer), so we should convert it to Long (longValue())

File: src/main/java/apoc/convert/Convert.java [hunk 2]
```diff
@@ -163,5 +165,85 @@ else if (list.getClass().isArray() && !list.getClass().getComponentType().isPrim
 	public List<Relationship> toRelationshipList(@Name("list") Object list) {
         return convertToList(list, Relationship.class);
 	}
-	
+
+    @UserFunction
+    @Description("apoc.convert.toInteger(value) | tries it's best to convert the value to an integer")
+    public Number toInteger(@Name("object") Object obj) {
```
Comment (jexp) [inline src/main/java/apoc/convert/Convert.java]: Should return Long

File: src/main/java/apoc/convert/Convert.java [hunk 3]
```diff
@@ -163,5 +165,85 @@ else if (list.getClass().isArray() && !list.getClass().getComponentType().isPrim
...
+            case BOOLEAN:
+                result = ((boolean) obj) ? 1 : 0;
+                break;
+            default:
+                throw new RuntimeException("Supported types are: Integer, Float, String, Boolean, HexString");
+        }
+        return result;
+    }
+
+    private Integer parseIntString(String input) {
```
Comment (jexp) [inline src/main/java/apoc/convert/Convert.java]: Should return long

File: src/main/java/apoc/convert/Convert.java [hunk 4]
```diff
@@ -163,5 +165,85 @@ else if (list.getClass().isArray() && !list.getClass().getComponentType().isPrim
...
+        if (input.equals("true")) {
+            return 1;
+        } else if (input.equals("false")) {
+            return 0;
+        } else if (input.startsWith("0x")) {
+            return Integer.valueOf(input.substring(2), 16);
```
Comment (jexp) [inline src/main/java/apoc/convert/Convert.java]: Long.valueOf()

File: src/main/java/apoc/convert/Convert.java [hunk 5]
```diff
@@ -163,5 +165,85 @@ else if (list.getClass().isArray() && !list.getClass().getComponentType().isPrim
...
+        } else {
+            return (int) Float.parseFloat(input);
```
Comment (jexp) [inline src/main/java/apoc/convert/Convert.java]: Shouldn't we have separate parsers for Long.parseLong() and (long)Double.parseDouble() based on existence of decimal point?

File: src/main/java/apoc/convert/Convert.java [hunk 6]
```diff
@@ -163,5 +165,85 @@ else if (list.getClass().isArray() && !list.getClass().getComponentType().isPrim
...
+    public Double toDouble(@Name("object") Object obj) {
+        if (obj == null || obj.equals("")) {
+            return null;
+        }
+
+        Types varType = Types.of(obj);
+        Double result;
+        switch (varType) {
+            case INTEGER:
+                result = ((Number) obj).doubleValue();
```
Comment (jexp) [inline src/main/java/apoc/convert/Convert.java]: just return without break

File: src/main/java/apoc/convert/Convert.java (hunk 1 above)
Comment (jexp) [inline src/main/java/apoc/convert/Convert.java]: just return without break

File: src/main/java/apoc/convert/Convert.java [hunk 7]
```diff
@@ -163,5 +165,85 @@ else if (list.getClass().isArray() && !list.getClass().getComponentType().isPrim
...
+            default:
+                throw new RuntimeException("Supported types are: Integer, Float, String, Boolean, HexString");
+        }
+        return result;
+    }
+
+    private Integer parseIntString(String input) {
+        if (input.equals("true")) {
+            return 1;
+        } else if (input.equals("false")) {
```
Comment (jexp) [inline src/main/java/apoc/convert/Convert.java]: no need for else if you return before.

File: src/main/java/apoc/convert/Convert.java (hunk 5 above)
Comment (zygm0nt) [reply src/main/java/apoc/convert/Convert.java]: There's a test for that:

'''
testCall(db, "return apoc.convert.toInteger('123') as value", r -> assertEquals(123L, r.get("value")));
        testCall(db, "return apoc.convert.toInteger('123.15') as value", r -> assertEquals(123L, r.get("value")));
'''

and it seems it's not necessary to differentiate based decimal point.

**Comment by zygm0nt** [comment]
@jexp fixed code according to your comments.

File: src/main/java/apoc/convert/Convert.java [hunk 8]
```diff
@@ -163,5 +165,76 @@ else if (list.getClass().isArray() && !list.getClass().getComponentType().isPrim
...
+        switch (varType) {
+            case INTEGER:
+            case FLOAT:
+                return ((Number) obj).longValue();
+            case STRING:
+                return parseLongString((String)obj);
+            case BOOLEAN:
+                return ((boolean) obj) ? 1L : 0L;
+            default:
+                throw new RuntimeException("Supported types are: Integer, Float, String, Boolean, HexString");
```
Comment (jexp) [inline src/main/java/apoc/convert/Convert.java]: return null

File: src/main/java/apoc/convert/Convert.java [hunk 9]
```diff
@@ -163,5 +165,76 @@ else if (list.getClass().isArray() && !list.getClass().getComponentType().isPrim
...
+        }
+
+        Types varType = Types.of(obj);
+        switch (varType) {
+            case INTEGER:
+                return ((Number) obj).doubleValue();
+            case STRING:
+                return parseDoubleString((String)obj);
+            case FLOAT:
+                return (Double) obj;
```
Comment (jexp) [inline src/main/java/apoc/convert/Convert.java]: ((Number)).doubleValue()

File: src/main/java/apoc/convert/Convert.java [hunk 10]
```diff
@@ -163,5 +165,76 @@ else if (list.getClass().isArray() && !list.getClass().getComponentType().isPrim
...
+            case BOOLEAN:
+                return ((boolean) obj) ? 1.0 : 0.0;
+            default:
+                throw new RuntimeException("Supported types are: Integer, Float, String, Boolean, HexString");
```
Comment (jexp) [inline src/main/java/apoc/convert/Convert.java]: return null

File: src/main/java/apoc/convert/Convert.java [hunk 11]
```diff
@@ -163,5 +165,76 @@ else if (list.getClass().isArray() && !list.getClass().getComponentType().isPrim
...
+        }
+    }
+
+
+    private Double parseDoubleString(String input) {
+        if (input.equals("true")) {
```
Comment (jexp) [inline src/main/java/apoc/convert/Convert.java]: equalsIgnoreCase

File: src/main/java/apoc/convert/Convert.java [hunk 12]
```diff
@@ -163,5 +165,76 @@ else if (list.getClass().isArray() && !list.getClass().getComponentType().isPrim
...
+            return 1.0;
+        }
+        if (input.equals("false")) {
+            return 0.0;
+        }
+        if (input.startsWith("0x")) {
+            Long i = Long.parseLong(input.substring(2), 16);
+            return Double.longBitsToDouble(i);
+        }
+        return Double.parseDouble(input);
```
Comment (jexp) [inline src/main/java/apoc/convert/Convert.java]: try catch else return null

File: src/main/java/apoc/convert/Convert.java [hunk 13]
```diff
@@ -163,5 +165,76 @@ else if (list.getClass().isArray() && !list.getClass().getComponentType().isPrim
...
+
+    private Double parseDoubleString(String input) {
+        if (input.equals("true")) {
+            return 1.0;
+        }
+        if (input.equals("false")) {
+            return 0.0;
+        }
+        if (input.startsWith("0x")) {
+            Long i = Long.parseLong(input.substring(2), 16);
```
Comment (jexp) [inline src/main/java/apoc/convert/Convert.java]: try catch

File: src/main/java/apoc/convert/Convert.java [hunk 14]
```diff
@@ -163,5 +165,76 @@ else if (list.getClass().isArray() && !list.getClass().getComponentType().isPrim
...
-	
+
+    @UserFunction
+    @Description("apoc.convert.toInteger(value) | tries it's best to convert the value to an integer")
+    public Long toInteger(@Name("object") Object obj) {
+        if (obj == null || obj.equals("")) {
+            return null;
+        }
+
+        Types varType = Types.of(obj);
```
Comment (jexp) [inline src/main/java/apoc/convert/Convert.java]: you can also add teh try-catch here, and return null in case of exception

**Comment by jexp** [comment]
ping @zygm0nt

**Comment by zygm0nt** [comment]
I've updated this PR. Please review @jexp

File: src/main/java/apoc/convert/Convert.java (hunk 14 above)
Comment (zygm0nt) [reply src/main/java/apoc/convert/Convert.java]: This seems pointless - Types.of() implementation returns UNKNOWN type as a fallback. It does not throw any exceptions AFAIK.

**Comment by jexp** [comment]
@zygm0nt there are some conflicts now, would you mind rebasing?

**Comment by jexp** [comment]
@zygm0nt ping?

**Comment by jexp** [comment]
merged manually

**Pull request closed without merging**